### PR TITLE
[WIP] Install missing dependencies if necessary

### LINKED ISSUE #11: Enhancement: Install flow-cli (runtime dependencies) if not found
### Issue To Be Solved

When the extension is published on the public Marketplace, users could download and install it without having first installed the flow-cli. The extension does not work without the cli already installed on a user's system.

### Possible Solution

Add "runtimeDependencies" to the extension config, and code for detecting and installing the flow-cli, if not detected. 
Similar to:
https://blog.lextudio.com/how-to-acquire-dependencies-for-visual-studio-code-extensions-9ca828cb168d

### PR DESCRIPTION
**Pull request by 10thfloor**
Closes #11

## Description
Adds code to check if user has the `flow-cli` installed, and installs it if not.

## TODO
- [ ] Windows support
- [x] Dynamic version number for Cadence
- [ ] Check if newer version is available and prompt for re-download
- [ ] Update extension state 
- [ ] Handle malformed request for the binary
- [ ] Show error messages / info messages where appropriate
- [x] Ensure downloads are cleaned up properly

## Nice to Have
- [ ] Support switching between multiple versions

### CODE REVIEW

File: src/flow-installer.ts [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+  }
+
+  async installDepsIfNotPresent() {
+    try {
+      await this.parsePlatform();
+      await this.checkForInstalledBin();
+      await this.checkForDownloadInProgress();
+      await this.startDownload();
+    } catch (e) {
+      console.log(e)
```
Comment (10thfloor) [inline src/flow-installer.ts:64]: Handle errors and info messages here.

File: src/flow-installer.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+  private async startDownload() {
+    switch (this.platform) {
+      case "win32":
+      // TODO
+      case "darwin":
+      case "linux":
+        if (fs.existsSync(this.downloadDest)) {
+          fs.unlinkSync(this.downloadDest);
+        }
```
Comment (10thfloor) [inline src/flow-installer.ts:122]:
```suggestion
```

File: src/flow-installer.ts [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+const DOWNLOAD_IN_PROGRESS = "Download in progress."
+const DOWNLOAD_FAILED = "Download failed."
+
+const DARWIN_FLOW_BIN_PATH = "/usr/local/bin"
+const LINUX_FLOW_BIN_PATH = `${process.env.HOME || process.env.HOMEPATH || process.env.USERPROFILE}/.local/bin`
+const WINDOWS_FLOW_BIN_PATH = ""
+
+/*
+ * Checks if the executable is on the PATH
+ */
```
Comment (10thfloor) [inline src/flow-installer.ts:27]:
```suggestion
```

File: src/flow-installer.ts [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+          if (this.inFlightDownloads.has(this.downloadURL)) {
+            this.inFlightDownloads.get(this.downloadURL)?.set(this.targetPath, downloadTask);
+          } else {
+            this.inFlightDownloads.set(this.downloadURL, new Map([[this.targetPath, downloadTask]]));
+          }
+          return downloadTask;
+        } catch (e) {
+          fs.unlinkSync(this.downloadDest);
+          throw new Error(`Failed to download ${url}`);
+        }
```
Comment (10thfloor) [inline src/flow-installer.ts:180]: Not sure if all this ceremony is necessary as "deduping" the call might not be an issue.

File: src/flow-installer.ts [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+  }
+
+
+
+  private async checkForInstalledBin() {
+    if (executableExists('flow')) {
+      throw <InfoMessage>{ message: BINARY_ALREADY_INSTALLED }
+    }
+    const downloadDest = this.downloadDest;
+    if (fs.existsSync(downloadDest)) {
```
Comment (10thfloor) [inline src/flow-installer.ts:99]: Remove partial download.

**Comment by sideninja** [comment]
Closing, as it is outdated and not needed anymore.

**Pull request closed without merging**